### PR TITLE
*: optimize memory usage

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve?

Issue Number: Ref #7897

### What is changed and how does it work?

```commit-message
```

### Check List

Tests

- Manual test 
compared with beb91c139
```
goos: linux
goarch: amd64
pkg: github.com/tikv/pd/server/cluster
cpu: Intel(R) Xeon(R) Gold 6240 CPU @ 2.60GHz
                         │   old.txt   │               new.txt                │
                         │   sec/op    │    sec/op     vs base                │
HandleRegionHeartbeat-72   5.049µ ± 9%   3.248µ ± 15%  -35.67% (p=0.000 n=10)

                         │  old.txt   │              new.txt               │
                         │    B/op    │    B/op     vs base                │
HandleRegionHeartbeat-72   989.5 ± 3%   571.0 ± 4%  -42.29% (p=0.000 n=10)

                         │  old.txt   │              new.txt               │
                         │ allocs/op  │ allocs/op   vs base                │
HandleRegionHeartbeat-72   25.00 ± 0%   14.00 ± 7%  -44.00% (p=0.000 n=10)
```

### Release note

```release-note
None.
```

### CODE REVIEW

File: pkg/ratelimit/runner.go [hunk 1]
```diff
@@ -61,7 +87,7 @@ type ConcurrentRunner struct {
 	limiter            *ConcurrencyLimiter
 	maxPendingDuration time.Duration
 	taskChan           chan *Task
-	pendingTasks       []*Task
+	pendingTasks       *list.List
```
Comment (nolouch) [inline pkg/ratelimit/runner.go:90]: why need a list?

**Review by nolouch — APPROVED** [review]

File: pkg/utils/grpcutil/grpcutil.go [hunk 2]
```diff
@@ -186,13 +186,9 @@ func ResetForwardContext(ctx context.Context) context.Context {
...
 // GetForwardedHost returns the forwarded host in metadata.
 func GetForwardedHost(ctx context.Context) string {
-	md, ok := metadata.FromIncomingContext(ctx)
-	if !ok {
-		log.Debug("failed to get gRPC incoming metadata when getting forwarded host")
-		return ""
-	}
-	if t, ok := md[ForwardMetadataKey]; ok {
-		return t[0]
+	s := metadata.ValueFromIncomingContext(ctx, ForwardMetadataKey)
```
Comment (HuSharp) [inline pkg/utils/grpcutil/grpcutil.go:189]: Do we need to change other `FromIncomingContext`?

Comment (rleungx) [reply pkg/utils/grpcutil/grpcutil.go:189]: I prefer not for now since it's still in experimental state.

File: pkg/ratelimit/runner.go [hunk 3]
```diff
@@ -197,7 +202,7 @@ func (cr *ConcurrentRunner) RunTask(ctx context.Context, f func(context.Context)
 		}
 		task.submittedAt = time.Now()
 		cr.pendingTasks = append(cr.pendingTasks, task)
-		cr.pendingTaskCount[taskOpts.TaskName]++
```
Comment (HuSharp) [inline pkg/ratelimit/runner.go:200]: Is `taskName` a required field? If so, should it be in the `task` struct?

File: server/cluster/cluster_test.go [hunk 4]
```diff
@@ -3754,11 +3755,72 @@ func BenchmarkHandleStatsAsync(b *testing.B) {
...
-
 	// Reset timer after setup
 	b.ResetTimer()
 	// Run HandleStatsAsync b.N times
 	for i := 0; i < b.N; i++ {
 		cluster.HandleStatsAsync(c, region)
 	}
 }
+
+func BenchmarkHandleRegionHeartbeat(b *testing.B) {
```
Comment (nolouch) [inline server/cluster/cluster_test.go:3766]: maybe we can run some benchmarks daily.

**Review by nolouch — APPROVED** [review]
lgtm

**Review by HuSharp — APPROVED** [review]

**Pull request merged**